File: public/metodeknn.js
// Membaca data dari file JSON
const jsonData = JSON.parse(fs.readFileSync('C:/skripsi/public/2minggu.json'));

// Mengubah data waktu menjadi detik sejak tengah malam
const habitData = [];
for (const key in jsonData.Data) {
  if (jsonData.Data.hasOwnProperty(key)) {
    const value = jsonData.Data[key];
    const waktu = new Date(`1970-01-01T${value.Waktu}Z`);
    const waktuDetik = waktu.getHours() * 3600 + waktu.getMinutes() * 60 + waktu.getSeconds();
    habitData.push([
      waktuDetik,
      value.dapur === 'hidup' ? 1 : 0,
      value.kamarUtama === 'hidup' ? 1 : 0,
      value.kamarKedua === 'hidup' ? 1 : 0,
      value.ruangTamu === 'hidup' ? 1 : 0,
      value.teras === 'hidup' ? 1 : 0,
      value.toilet === 'hidup' ? 1 : 0
    ]);
  }
}

// Fungsi untuk menghitung jarak Euclidean hanya berdasarkan waktu
function euclideanDistance(x1, x2) {
  return Math.abs(x1 - x2);
}

// Fungsi untuk memprediksi keadaan lampu berdasarkan waktu terdekat
function predictLightsByTime(time, k) {
  const waktu = new Date(`1970-01-01T${time}Z`);
  const waktuDetik = waktu.getHours() * 3600 + waktu.getMinutes() * 60 + waktu.getSeconds();
  const habits = [
    waktuDetik,
    ...(jsonData.Data['-NUbhKQusYWmh4qcwaEW'] ? [jsonData.Data['-NUbhKQusYWmh4qcwaEW'].dapur === 'hidup' ? 1 : 0,
    jsonData.Data['-NUbhKQusYWmh4qcwaEW'].kamarUtama === 'hidup' ? 1 : 0,
    jsonData.Data['-NUbhKQusYWmh4qcwaEW'].kamarKedua === 'hidup' ? 1 : 0,
    jsonData.Data['-NUbhKQusYWmh4qcwaEW'].ruangTamu === 'hidup' ? 1 : 0,
    jsonData.Data['-NUbhKQusYWmh4qcwaEW'].teras === 'hidup' ? 1 : 0,
    jsonData.Data['-NUbhKQusYWmh4qcwaEW'].toilet === 'hidup' ? 1 : 0
    ] : [0, 0, 0, 0, 0, 0])
  ];

  // Menghitung jarak Euclidean hanya berdasarkan waktu dengan semua data dalam habitData
  const distances = habitData.map(habit => euclideanDistance(habit[0], habits[0]));

  // Mengambil indeks k tetangga terdekat
  const kIndices = Array.from({ length: k }, (_, i) => i)
    .sort((a, b) => distances[a] - distances[b]);

  // Mengambil status lampu dari k tetangga terdekat
  const neighborLights = kIndices.map(index => habitData[index].slice(1));

  const nNeighbors = kIndices.length;
  const predictedLights = [];
  for (let i = 0; i < neighborLights[0].length; i++) {
    const lamp = neighborLights.map(neighbor => neighbor[i]);
    const lampState = lamp.reduce((acc, val) => acc + val, 0) > nNeighbors / 2 ? 1 : 0;
    predictedLights.push(lampState);
  }

  return predictedLights;
}

// Masukkan waktu baru di sini
// const waktuTerbaru = new Date();
// const formattedTime = waktuTerbaru.toISOString().substr(11, 8);
// const newTime = formattedTime; // Contoh waktu baru
// const kNeighbors = 3;

const predictedLights = predictLightsByTime(newTime, kNeighbors);
console.log("Hasil prediksi keadaan lampu (0: mati, 1: hidup):");
console.log(predictedLights);

const actualData = jsonData.Data;

const predictedData = {};
for (const key in actualData) {
  if (actualData.hasOwnProperty(key)) {
    const value = actualData[key];
    const waktu = value.Waktu;
    const predictedLights = predictLightsByTime(waktu, kNeighbors);
    predictedData[key] = predictedLights;
  }
}

// Inisialisasi confusion matrix
const confusionMatrix = {
  truePositive: 0,
  falsePositive: 0,
  trueNegative: 0,
  falseNegative: 0
};

for (const key in actualData) {
  if (actualData.hasOwnProperty(key)) {
    const value = actualData[key];
    const actualLights = [
      value.dapur === 'hidup' ? 1 : 0,
      value.kamarUtama === 'hidup' ? 1 : 0,
      value.kamarKedua === 'hidup' ? 1 : 0,
      value.ruangTamu === 'hidup' ? 1 : 0,
      value.teras === 'hidup' ? 1 : 0,
      value.toilet === 'hidup' ? 1 : 0
    ];
    const predictedLights = predictedData[key];

    for (let i = 0; i < actualLights.length; i++) {
      if (actualLights[i] === 1 && predictedLights[i] === 1) {
        confusionMatrix.truePositive++;
      } else if (actualLights[i] === 0 && predictedLights[i] === 1) {
        confusionMatrix.falsePositive++;
      } else if (actualLights[i] === 0 && predictedLights[i] === 0) {
        confusionMatrix.trueNegative++;
      } else if (actualLights[i] === 1 && predictedLights[i] === 0) {
        confusionMatrix.falseNegative++;
      }
    }
  }
}

const accuracy = (confusionMatrix.truePositive + confusionMatrix.trueNegative) / (confusionMatrix.truePositive + confusionMatrix.falsePositive + confusionMatrix.trueNegative + confusionMatrix.falseNegative);
console.log("Tingkat Akurasi:", accuracy);

const predictButton = document.getElementById("predictButton");
const resultDiv = document.getElementById("result");

predictButton.addEventListener("click", () => {
  const waktuTerbaru = new Date();
  const formattedTime = waktuTerbaru.toISOString().substr(11, 8);
  const newTime = formattedTime; // Contoh waktu baru
  const kNeighbors = 3;

  const predictedLights = predictLightsByTime(newTime, kNeighbors);

  // Tampilkan hasil prediksi di dalam elemen HTML
  resultDiv.innerHTML = "Hasil prediksi keadaan lampu (0: mati, 1: hidup):<br>" + predictedLights.join(", ");
});

File: public/metodeKNN.py
import json
import numpy as np
from datetime import datetime

# Membaca data dari file JSON
with open('C:/skripsi/public/2minggu.json') as json_file:
    data = json.load(json_file)

# Mengubah data waktu menjadi detik sejak tengah malam
habit_data = []
for key, value in data['Data'].items():
    waktu = datetime.strptime(value['Waktu'], "%H:%M:%S")
    waktu_detik = waktu.hour * 3600 + waktu.minute * 60 + waktu.second
    habit_data.append([
        waktu_detik,
        1 if value['dapur'] == 'hidup' else 0,
        1 if value['kamar'] == 'hidup' else 0,
        1 if value['kamar2'] == 'hidup' else 0,
        1 if value['ruangtamu'] == 'hidup' else 0,
        1 if value['teras'] == 'hidup' else 0,
        1 if value['toilet'] == 'hidup' else 0
    ])

habit_data = np.array(habit_data)



# Fungsi untuk menghitung jarak Euclidean hanya berdasarkan waktu
def euclidean_distance(x1, x2):
    return np.abs(x1 - x2)

# Fungsi untuk memprediksi keadaan lampu berdasarkan waktu terdekat
def predict_lights_by_time(time, k):
    waktu = datetime.strptime(time, "%H:%M:%S")
    waktu_detik = waktu.hour * 3600 + waktu.minute * 60 + waktu.second
    habits = np.array([
        waktu_detik,
        *[1 if value == 'hidup' else 0 for value in data['Data']['-NUbhKQusYWmh4qcwaEW'].values()]
    ])

    # Menghitung jarak Euclidean hanya berdasarkan waktu dengan semua data dalam habit_data
    distances = [euclidean_distance(habits[0], habit[0]) for habit in habit_data]

    # Mengambil indeks k tetangga terdekat
    k_indices = np.argsort(distances)[:k]

    # Mengambil status lampu dari k tetangga terdekat
    neighbor_lights = habit_data[k_indices][:, 1:]

    n_neighbors = len(k_indices)
    predicted_lights = []
    for lamp in neighbor_lights.T:
        lamp_state = 1 if np.sum(lamp) > n_neighbors / 2 else 0
        predicted_lights.append(lamp_state)

    return predicted_lights

# Masukkan waktu baru di sini
waktu_terbaru = datetime.now()
formatted_time = waktu_terbaru.strftime("%H:%M:%S")
new_time = formatted_time  # Contoh waktu baru
k_neighbors =21

predicted_lights = predict_lights_by_time(new_time, k_neighbors)
print("Hasil prediksi keadaan lampu (0: mati, 1: hidup):")
print(predicted_lights)

with open('C:/skripsi/public/2minggu.json') as json_file:
    actual_data = json.load(json_file)['Data']

predicted_data = {}
for key, value in actual_data.items():
    waktu = value['Waktu']
    predicted_lights = predict_lights_by_time(waktu, k_neighbors)
    predicted_data[key] = predicted_lights

    # Inisialisasi confusion matrix
confusion_matrix = {
    'true_positive': 0,
    'false_positive': 0,
    'true_negative': 0,
    'false_negative': 0,
}
print(confusion_matrix)

for key, value in actual_data.items():
    actual_lights = [
        1 if value['dapur'] == 'hidup' else 0,
        1 if value['kamar'] == 'hidup' else 0,
        1 if value['kamar2'] == 'hidup' else 0,
        1 if value['ruangtamu'] == 'hidup' else 0,
        1 if value['teras'] == 'hidup' else 0,
        1 if value['toilet'] == 'hidup' else 0
    ]
    predicted_lights = predicted_data[key]


    for i in range(len(actual_lights)):
        if actual_lights[i] == 1 and predicted_lights[i] == 1:
            confusion_matrix['true_positive'] += 1
        elif actual_lights[i] == 0 and predicted_lights[i] == 1:
            confusion_matrix['false_positive'] += 1
        elif actual_lights[i] == 0 and predicted_lights[i] == 0:
            confusion_matrix['true_negative'] += 1
        elif actual_lights[i] == 1 and predicted_lights[i] == 0:
            confusion_matrix['false_negative'] += 1


print(confusion_matrix['true_positive'])
print(confusion_matrix['false_positive'])
print(confusion_matrix['true_negative'])
print(confusion_matrix['false_negative'])
accuracy = (confusion_matrix['true_positive'] + confusion_matrix['true_negative']) / (confusion_matrix['true_positive'] + confusion_matrix['false_positive'] + confusion_matrix['true_negative'] + confusion_matrix['false_negative'])
print("Tingkat Akurasi:", accuracy * 100, "%")